### DISCUSSION TITLE
Breaking Change with PHP 7.2/8 on Dreamhost?

### DISCUSSION THREAD

**Discussion by irthinker** [PHPMailer/PHPMailer#2976] [General]
I have had PHPMailer working beautifully on my site for many years. I upgraded to PHP 8.1 in the past 3-6 months and I honestly can't recall what version of PHP I was on. My existing PHP Mailer files are from 8/29/2023 and I *think* the mail functions were working. To be honest, they are only used to create accounts and reset passwords for the most part, so not as obvious a problem. I am sure they were working in February of this year, because usage was hot and heavy then.

Anyway, right now, my code (which, again, was working for years) breaks when I hit this line:

$mail = new PHPMailer;

I am not entirely sure how to debug that. Any help would be appreciated.

Is it a PHP upgrade thing? A Dreamhost change thing?

**Comment by Synchro** [comment]
Um, did you not read the readme or any of the examples that show how to import PHPMailer classes into the global namespace? Add either a `use` statement or use a fully qualified class name. Reading the upgrade guide might help too.

**Comment by irthinker** [comment]
Marcus, Yes. I am a big fan of PHP Mailer and I have been using PHPMailer
for longer than I can remember. I also succesffull took it through the
namespace upgrade. What really inspired me to make this general post was
that for some reason the code that has been working breaks at that very
basic initial step and things like the known GoDaddy issue prompted me to
take a shot in the dark. I was hoping someone would come back and say, "Ah,
yes, Michael, Dreamhost did X and you just need to do Y. Problem solved!"
or "Ah, yes, in PHP 8, this weird little bit that they got picky about may
cause that."

Because the page fails at that very first step, I can't even get to the
point where it will see my DEBUG setting.

I may just have to nuke it from orbit and start from scratch.

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Wed, Nov 15, 2023 at 3:01 AM Marcus Bointon ***@***.***>
wrote:

> Um, did you not read the readme or any of the examples that show how to
> import PHPMailer classes into the global namespace? Add either a use
> statement or use a fully qualified class name. Reading the upgrade guide
> might help too.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PHPMailer/PHPMailer/discussions/2976#discussioncomment-7574491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHDZVNTP5O7VLPI4TLS24LYESAHNAVCNFSM6AAAAAA7LX4CUWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TKNZUGQ4TC>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Michael

**Reply by Synchro** [reply]
One key problem here is that you say "it breaks", but don't say how. Syntax error? Class not found? We can only guess without an error message.

**Comment by irthinker** [comment]
My apologies.

I was using "echo" as a smoke test.

I slid one in before and after. The one before executed, the one after did
not.

My problem is that my testers apparently did virtually no testing. I and
they are volunteers with limited time to work on this and the mail feature
is only invoked for very specific things (account creation and
verification, password reset, and some notifications), so it's quite
understandable that it slipped by.

It's being hit heavily now so I was loath to switch on errors, but I will
try to look into that after I get off work and before my other volunteer
work starts.

Apologies for the sub-optimal context.

Michael

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Wed, Nov 15, 2023 at 9:01 AM Marcus Bointon ***@***.***>
wrote:

> One key problem here is that you say "it breaks", but don't say how.
> Syntax error? Class not found? We can only guess without an error message.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PHPMailer/PHPMailer/discussions/2976#discussioncomment-7578081>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHDZVM2BJBI36P2JUC7K3TYETKNXAVCNFSM6AAAAAA7LX4CUWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TKNZYGA4DC>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Michael

**Reply by Synchro** [reply]
It shouldn't be a matter of "switching on" errors – you should find any PHP errors in your web server's error log, or on the console if you run the script manually. If you can add 'echo' to a script, you can temporarily make errors very visible by adding `ini_set('display_errors', 1);`.

**Comment by irthinker** [comment]
Ok, thanks for the clarification.

I found a minute to add that and run it:

" *Fatal error*: Uncaught Error: Class "PHPMailer" not found"

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Wed, Nov 15, 2023 at 12:25 PM Marcus Bointon ***@***.***>
wrote:

> It shouldn't be a matter of "switching on" errors – you should find any
> PHP errors in your web server's error log, or on the console if you run the
> script manually. If you can add 'echo' to a script, you can temporarily
> make errors very visible by adding ini_set('display_errors', 1);.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PHPMailer/PHPMailer/discussions/2976#discussioncomment-7580131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHDZVMKR5DKBFF663JGUC3YEUCKZAVCNFSM6AAAAAA7LX4CUWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TKOBQGEZTC>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Michael

**Reply by Synchro** [reply]
This is what I suspected. This means one of two things: either the class is not loaded, or you have not imported it into the namespace. Change your line of code to this:

    $mail = new \PHPMailer\PHPMailer\PHPMailer();

Then run it again. If you get the same error, it will be because the class is not loaded, in which case you need to check that your composer setup is working, or if you're not using that, that you have appropriate `require` statements to load the classes. See the readme for [more info on loading PHPMailer](https://github.com/PHPMailer/PHPMailer#installation--loading).

**Comment by irthinker** [comment]
Sounds good, thank you for your help and thanks for PHP Mailer!

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Wed, Nov 15, 2023 at 2:00 PM Marcus Bointon ***@***.***>
wrote:

> This is what I suspected. This means one of two things: either the class
> is not loaded, or you have not imported it into the namespace. Change your
> line of code to this:
>
> $mail = new \PHPMailer\PHPMailer\PHPMailer();
>
> Then run it again. If you get the same error, it will be because the class
> is not loaded, in which case you need to check that your composer setup is
> working, or if you're not using that, that you have appropriate require
> statements to load the classes. See the readme for more info on that.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PHPMailer/PHPMailer/discussions/2976#discussioncomment-7580828>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHDZVP7UMEDQUICYYUYOATYEUNM5AVCNFSM6AAAAAA7LX4CUWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TKOBQHAZDQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Michael

**Reply by irthinker** [reply]
Thank you again for helping ID the core issue. I poked around and found that other mails were working, but they were all done within the page, whereas this one was through a method call. For this broken case, I pulled the code directly into the page and it worked just fine.

So, it seems that the issue is that, for some reason that is not clear to me, having the code in an Object Method was the problem.

Has there been a change this year related to name spaces and calling PHPMailer in a method that might compromise the ability to call PHPMailer within an object method?

**Reply by Synchro** [reply]
Nothing major in the last year. It's bound to be something simple, most likely that the PHPMailer class was not imported into the local namespace in the file it was breaking in. PHP `use` statements only apply to the file they appear in, and if you miss that, copying code into a new location will break that way. I can't be more precise without seeing the code.

**Reply by irthinker** [reply]
Understood, thanks. I am using an object in a sitewide library for creating pages and it contains a method for sending an email using PHPMailer. In the given page I have created from that I do apply use for the namespace in the page (as I do with any page using PHPMailer) and then invoke the method from the site library to send the email.

Is there an example of taking that approach in the docs? I was not able to find it, but I certainly could have missed it. Totally happy to self-serve if I can see one simple working example like that.

**Reply by irthinker** [reply]
Basically, I must be failing to do something important when I shove it into a method, since it works perfectly otherwise. :-)

**Reply by Synchro** [reply]
I'm flying blind here without seeing your code, but it should be very simple, along the lines of:

```php
<?php
use PHPMailer\PHPMailer\PHPMailer;

require '../vendor/autoload.php';

class thing
{
    public function doSomething()
    {
        $mail = new PHPMailer();
        //other stuff here...
    }
}
```

Basically, if you're missing the `use` statement, that `new PHPMailer()` call will fail. This code also assumes you're using composer to load the classes (which is highly recommended), but if you're not, replace that line with `require '/path/to/PHPMailer/src/PHPMailer.php';`

**Reply by irthinker** [reply]
I hear you. Thank you for patience and generosity.

To frame it in that sense with the relevant lines of code:

In the page I am calling from, I have:

```php
<?php

use PHPMailer\PHPMailer\PHPMailer;
use PHPMailer\PHPMailer\SMTP;
use PHPMailer\PHPMailer\Exception;`

$this_page = new pageBase;

$this_page->main_body .= $this_page->sendVerificationEmail($email_address, $verification_needed, $verification_lookup);
```

And in the library, I have:

```php
<?php

class pageBase {

public function sendVerificationEmail($email_address, $verification_needed, $verification_lookup) {

        require_once 'src/phpmailer/Exception.php';
        require_once 'src/phpmailer/PHPMailer.php';
        require_once 'src/phpmailer/SMTP.php';
        $mail = new PHPMailer;
```

Obviously, a lot more code is involved, but based on the namespace issue we think we are diagnosing, I am guessing the above should be where the problem lies.

And, again, if I replace the method call with the code from the method and replace $this-> with $this_page, it works just fine directly in the page.

Thoughts?

**Reply by Synchro** [reply]
As I suspected, *you're missing the `use` statements* from your library file, and you don't need them in the page script. They only work *in the file they are present in*, so if you use a namespaced class like PHPMailer, you need to import it into the current namespace in *every file you use it in*. Note that it is pretty much independent of *loading* the classes: `use` means "refer to this class using this name", but `require` means "load a PHP script from this file"; they are quite different things.

**Reply by irthinker** [reply]
Fantastic! I added the three lines to library file and voila! It works as expected. Merci!
